### PR TITLE
Feature: Add custom CalendarDatePicker date parsing to allow for multiple user input formats

### LINKED ISSUE #21179: Feature request: Easily allow overriding date parsing in CalendarDatePicker
### Is your feature request related to a problem? Please describe.

I allow my user to choose the date format for a calendar date picker's input (`CustomDateFormatString` is determined by the user, essentially). This works fine.

However, when the user _types_ in a new date, they are forced to type it in the custom format that is being displayed. 

This behavior cannot be easily overridden as the text parsing method is private: 

https://github.com/AvaloniaUI/Avalonia/blob/b7e95c2b0961c33f06a3a64846c4207fb406eada/src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.cs#L714-L724

### Describe the solution you'd like

I would like to control from my app the valid date formats that the user could type in with a fallback to the standard behavior if my custom parsing logic does not accept the given text input. For example, if the app is using `yyyy-MM-dd` as the date, the user should also be able to type dates like `Apr 13 2025` or `13 Apr 2025` or `13 April 2025` or similar, and it all gets parsed OK despite those inputs being different from the display format. (As it stands right now, if using `yyyy-MM-dd` for display, they can only type in a date with the format `yyyy-MM-dd`.)

I think one, simple way to move forward would be to make that method `protected virtual` so a subclass can just override it, but this may have a perf impact, I'm guessing. Another method would be some `Func` or similar that could be called by `ParseText`, as suggested here a few years ago: https://github.com/AvaloniaUI/Avalonia/discussions/13464#discussioncomment-7458312

### Describe alternatives you've considered

* https://github.com/AvaloniaUI/Avalonia/discussions/13464 -- There seems to be some way to hack into the `TextBox` itself to override the text input that I'm not quite following, but this seems less than ideal.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Deadpikle**
## What does the pull request do?

Fixes #21179 by adding a `TextConverter` property to `CalendarDatePicker` so that the user can allow for multiple date formats for user entry.

## What is the current behavior?

When using `CalendarDatePickerFormat.Custom` and `CustomDateFormatString`, only dates in the `CustomDateFormatString` format can be typed in by the user.

## What is the updated/expected behavior with this PR?

Developers can now allow for multiple date formats being entered by the user by implementing the `CalendarDatePicker.TextConverter` property with a custom `IValueConverter`.

## How was the solution implemented (if it's not obvious)?

Copied the overall details from `NumericUpDown` per suggestion in #21179. Otherwise fairly standard.

Because users can now type in a different format than what is displayed, also tweaked the `ParseText` method to make sure the right format is always displayed to the user when the value is set.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

None

## Obsoletions / Deprecations

None

## Fixed issues

Fixes #21179

### CODE REVIEW

**Comment by cla-avalonia** [comment]
- [x] All contributors have signed the CLA.

**Comment by Deadpikle** [comment]
@cla-avalonia agree

File: src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.cs [hunk 1]
```diff
@@ -853,7 +860,9 @@ private void SetSelectedDate()
                 DateTime? d = ParseText(s);
                 if (d != null)
                 {
-                    SetValue(TextProperty, s);
+                    // make sure displayed text is reformatted to correct date format
+                    string? newtext = DateTimeToString((DateTime)d);
```
Comment (MrJul) [inline src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.cs:871]: `DateTimeToString` should be modified to call the converter (add matching tests).

File: src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.cs [hunk 2]
```diff
@@ -713,15 +713,22 @@ private void OpenDropDown()
...
             // TryParse is not used in order to be able to pass the exception to
             // the TextParseError event
             try
             {
-                newSelectedDate = SelectedDateFormat == CalendarDatePickerFormat.Custom && !string.IsNullOrEmpty(CustomDateFormatString) ?
-                    DateTime.ParseExact(text, CustomDateFormatString, DateTimeHelper.GetCurrentDateFormat()) :
-                    DateTime.Parse(text, DateTimeHelper.GetCurrentDateFormat());
+                if (TextConverter != null)
+                {
+                    newSelectedDate = (DateTime?)TextConverter.Convert(text, typeof(DateTime?), null, CultureInfo.CurrentCulture);
```
Comment (MrJul) [inline src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.cs]: `ConvertBack` (text to date) should be called instead of `Convert` (date to text).

File: src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.Properties.cs [hunk 3]
```diff
@@ -265,11 +274,28 @@ public CalendarDatePickerFormat SelectedDateFormat
...
         public string CustomDateFormatString
         {
             get => GetValue(CustomDateFormatStringProperty);
             set => SetValue(CustomDateFormatStringProperty, value);
         }
+        
+        /// <summary> 
+        /// Gets or sets the custom bidirectional Text-Value converter. 
+        /// </summary> 
+        public IValueConverter? TextConverter
```
Comment (MrJul) [inline src/Avalonia.Controls/CalendarDatePicker/CalendarDatePicker.Properties.cs:299]: Clarify that the property is only used with `Custom` formats and that it has priority over `CustomDateFormatString`.

**Review by MrJul — CHANGES_REQUESTED** [review]

**Comment by Deadpikle** [comment]
Thanks for reviewing! Code has been updated. Also added tests for what happens when user inputs invalid date format or sets `SelectedDate` to null.

**Review by MrJul — APPROVED** [review]
LGTM!
